### ISSUE TITLE
Remote current Temperature overwritten by local current temperature

### ISSUE THREAD

**Issue by victorgolfecho** [echavet/MitsubishiCN105ESPHome#204]
I setup a remote temperature sensor which is correctly updates the temperature based on the HA entity id.  However, the Current Temperate is regularly overwritten by the room temperature being provided by unit.  Note: this behavior is happening on only 1 of my 2 heat pumps (they are different models) configured with a remote thermostat.

I enabled DEBUG log and captured when it's happening.  In this example, the remote temp is reading 19.5C and the unit is reading 11C.

> [13:30:27][D][CN105:308]: sending a request for settings packet (0x02)
[13:30:27][I][CYCLE:039]: 1: Cycle start
[13:30:27][D][CYCLE:311]: 2a: Sending settings request (0x02)
[13:30:27][D][CN105:323]: creating Info packet
[13:30:27][D][CN105:084]: writing packet...
[13:30:27][D][WRITE:170]: FC 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7B 
[13:30:27][D][Header:091]: command: (62) data length: [10]<-- header
[13:30:27][D][READ:170]: FC 62 01 30 10 02 00 00 01 01 0F 02 05 00 00 00 A0 00 00 00 00 A3 
[13:30:27][D][chkSum:077]: OK-> A3=A3 
[13:30:27][D][CYCLE:315]: 2b: Receiving settings response
[13:30:27][D][Decoder:173]: [0x02 is settings]
[13:30:27][D][Decoder:181]: [Power : ON]
[13:30:27][D][Decoder:182]: [iSee  : 0]
[13:30:27][D][Decoder:183]: [Mode  : HEAT]
[13:30:27][D][Decoder:195]: [Temp °C: 16.000000]
[13:30:27][D][Decoder:198]: [Fan: 1]
[13:30:27][D][Decoder:201]: [Vane: ↓↓]
[13:30:27][W][lookup:204]: wideVane reading caution: value 0 not found, returning value at index 0
[13:30:27][D][Decoder:206]: [wideVane: ←← (adj:0)]
[13:30:27][D][SETTINGS:113]: [current        ]-> [power: ON , target °C: 16.0, mode: HEAT  , fan: 1     , vane: ↓↓, wvane: ←←]
[13:30:27][D][SETTINGS:113]: [received       ]-> [power: ON , target °C: 16.0, mode: HEAT  , fan: 1     , vane: ↓↓, wvane: ←←]
[13:30:27][D][EVT_SETS:072]: [wanted         ]-> [power: -  , target °C: -1.0, mode: -     , fan: -     , vane: -     , wvane: -     , hasChanged ? ->  NO, hasBeenSent ? ->  NO]
[13:30:27][D][SETTINGS:087]: [climate]-> [mode: HEAT, target °C: 16.0, fan: LOW, swing: OFF]
[13:30:27][D][CN105:218]: setting action to -> 4
[13:30:27][D][CN105:267]: Climate mode is: 3
[13:30:27][D][CN105:268]: Climate action is: 4
[13:30:27][D][climate:396]: 'Heat Pump - Living Room' - Sending state:
[13:30:27][D][climate:399]:   Mode: HEAT
[13:30:27][D][climate:401]:   Action: IDLE
[13:30:27][D][climate:404]:   Fan Mode: LOW
[13:30:27][D][climate:416]:   Swing Mode: OFF
[13:30:27][D][climate:419]:   Current Temperature: 19.50°C
[13:30:27][D][climate:425]:   Target Temperature: 16.00°C
[13:30:27][D][CYCLE:318]: 3a: Sending room °C request (0x03)
[13:30:27][D][CN105:323]: creating Info packet
[13:30:27][D][CN105:084]: writing packet...
[13:30:27][D][WRITE:170]: FC 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7A 
[13:30:27][W][component:237]: Component esphome.coroutine took a long time for an operation (75 ms).
[13:30:27][W][component:238]: Components should block for at most 30 ms.
[13:30:27][D][Header:091]: command: (62) data length: [10]<-- header
[13:30:27][D][READ:170]: FC 62 01 30 10 03 00 00 01 00 92 96 00 00 00 00 00 00 00 00 00 31 
[13:30:27][D][chkSum:077]: OK-> 31=31 
[13:30:27][D][CYCLE:324]: 3b: Receiving room °C response
[13:30:27][D][Decoder:253]: [Room °C: 11.000000]
[13:30:27][D][Decoder:254]: [OAT  °C: 9.000000]
[13:30:27][I][STATUS:134]: [received       ]-> [room C°: 11.0, operating: NO , compressor freq: 0.0 Hz]
[13:30:27][I][STATUS:134]: [current        ]-> [room C°: 19.5, operating: NO , compressor freq: 0.0 Hz]
[13:30:27][D][CN105:218]: setting action to -> 4
[13:30:27][D][CN105:267]: Climate mode is: 3
[13:30:27][D][CN105:268]: Climate action is: 4
[13:30:27][D][climate:396]: 'Heat Pump - Living Room' - Sending state:
[13:30:27][D][climate:399]:   Mode: HEAT
[13:30:27][D][climate:401]:   Action: IDLE
[13:30:27][D][climate:404]:   Fan Mode: LOW
[13:30:27][D][climate:416]:   Swing Mode: OFF
[13:30:27][D][climate:419]:   Current Temperature: 11.00°C
[13:30:27][D][climate:425]:   Target Temperature: 16.00°C
[13:30:27][D][CYCLE:327]: 4a: Sending status request (0x06)
[13:30:27][D][CN105:323]: creating Info packet
[13:30:27][D][CN105:084]: writing packet...
[13:30:27][D][WRITE:170]: FC 42 01 30 10 06 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 77 
[13:30:27][W][component:237]: Component esphome.coroutine took a long time for an operation (56 ms).
[13:30:27][W][component:238]: Components should block for at most 30 ms.
[13:30:27][D][Header:091]: command: (62) data length: [10]<-- header

You can see this line:
>[13:30:27][I][STATUS:134]: [received       ]-> [room C°: 11.0, operating: NO , compressor freq: 0.0 Hz]
[13:30:27][I][STATUS:134]: [current        ]-> [room C°: 19.5, operating: NO , compressor freq: 0.0 Hz]

is from here: https://github.com/echavet/MitsubishiCN105ESPHome/blob/89e525a51288245a5df82120980044d5ed53eb12/components/cn105/hp_readings.cpp#L432

which sets room temp on line 444.

Here is the relevant parts of the config yaml:

> sensor:
>   - platform: homeassistant
>     name: "Temperature Sensor From Home Assistant"
>     entity_id: ${remote_temp_sensor}    
>     internal: false
>     device_class: temperature
>     state_class: measurement
>     unit_of_measurement: "°C"
>     filters:
>       - lambda: 'return (x - 32) * (5.0/9.0);'
>       - clamp: # Limits values to range accepted by Mitsubishi units
>           min_value: 1
>           max_value: 40
>           ignore_out_of_range: true
>       - throttle: 30s
>     on_value:
>       then:
>         - logger.log:
>             level: WARN
>             format: "Remote temperature received from HA: %.1f C"
>             args: [ 'x' ]
>         - lambda: 'id(hp).set_remote_temperature(x);'

> climate:
>   - platform: cn105
>     name: ${friendly_name}
>     id: hp
>     remote_temperature_timeout: 30min
>     update_interval: 4s
>     debounce_delay : 100ms

Lastly, this is easily reproducible on my side.  It oscillates between the two temp readings every few seconds, happy to provide any more info if it'd be helpful.

**Comment by CopperSubstrate** [comment]
I've had this issue as well.

**Comment by victorgolfecho** [comment]
I have a change into the geoffdavis version of this project that appears to be working for me: https://github.com/geoffdavis/esphome-mitsubishiheatpump/pull/173

I have not had a chance to make the changes to this version.

**Comment by phidauex** [comment]
@victorgolfecho What indoor and outdoor unit models do you have? I wonder how the MHK and Kumo systems deal with the issue if something about the indoor unit overwrites the remote temperature value on its own (perhaps they do what your proposed solution does, and simply republish the remote temperature frequently).

**Comment by echavet** [comment]
> I have the same issue, it is updating from the remote sensor, but it's mixing up for some reason.
> 
> I also noticed my unit is not stopping when at target temperature, there seems to be a built-in offset of at least 1 degree.
> 
> 
> 
> @echavet any idea in which direction a solution can be found for this issue?

No. I need to dig!

**Comment by echavet** [comment]
Hi @victorgolfecho,
I'm very surprised by the magnitude of the difference between the two temperature probes.
How do you explain this discrepancy?
The bug is indeed visible in the logs, but I'd like to better understand the context in which it occurs.

**Comment by echavet** [comment]
hey everyone, 
could you provide your unit ref (which is facing the issue of course)?

**Comment by echavet** [comment]
> I have a change into the geoffdavis version of this project that appears to be working for me: [geoffdavis/esphome-mitsubishiheatpump#173](https://github.com/geoffdavis/esphome-mitsubishiheatpump/pull/173)
> 
> I have not had a chance to make the changes to this version.

@victorgolfecho 
it is very confusing for me.
The geoffrey version is very different. Do you mean you've had the same issue with geoff Davis's version, but you've got it fixed with this commit?

**Comment by jan-dewit** [comment]
> hey everyone, could you provide your unit ref (which is facing the issue of course)?

I'm using a MSZ-HR35VF.

**Comment by CopperSubstrate** [comment]
Two units seem impacted: a MSZ-GL18NA-U1 and a MSZ-GL06NA-U1

**Comment by youknowjack0** [comment]
I faced this issue with a SEZ-KD unit - i.e. it "forgets" remote temperature after about a minute. I resolved it simply by adding a 15s "heartbeat" filter to the remote sensor to keep republishing the remote temp, something like:

```
sensor:
- platform: homeassistant
  [...]
  filters:
    - heartbeat: 15s        
  [...]
  on_value:
      - lambda: 'id(hp).set_remote_temperature(x);'
```

**Comment by victorgolfecho** [comment]
@youknowjack0 heartbeat fix definitely works.  I validated it a few days ago and it's been perfect for the last 36 hours.  This is wonderful!

**Comment by tfasz** [comment]
I ran into this issue as well (my units are ducted so the "room" temp is the un-insulated attic/basement temp). I've applied the same work-around but it's a bit of a concern that it's just masking an issue/race condition.

**Comment by echavet** [comment]
I would like to move forward with resolving this issue, but I will need some help.  

Several hypotheses:  
- some air conditioning models might have a specific behavior regarding the update of the external temperature.  
- it might be a peculiarity of my implementation(It would be necessary to test with GeoffDavis' implementation to confirm this).  
- It is possible that the cycle duration (which is logged when the `CYCLE: DEBUG` logger is enabled) exceeds that of `update_interval`. The stress caused by sending packets to the air conditioner might make it skip processing the external temperature packet. You can try temporarily increasing `update_interval` to verify this hypothesis.  

I have added specific logs in the next firmware update (issue #245) to more accurately trace the calls to `set_remote_temperature()`.  

```yaml
logger:
  #hardware_uart: USB_CDC
  baud_rate: 0
  level: DEBUG
  logs:
    ACK : DEBUG
    REMOTE_TEMP : DEBUG
```
with this set, it is expected to read something like that:

```
[10:26:45][D][REMOTE_TEMP:301]: Sending remote temperature...
[10:26:45][D][REMOTE_TEMP:373]: Sending remote temperature packet... -> 20.799999
[10:26:45][I][CYCLE:052]: 6: Cycle ended in 1.0 seconds (with timeout?:  NO)
[10:26:45][D][Update-ACK:170]: FC 61 01 30 10 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 5E 
[10:26:45][D][ACK:407]: Last heatpump data update successful!
```

If you have any additional information, I’m interested.  

In any case, please note that the external temperature sensor value is always sent at the end of a cycle.  

```cpp
void CN105Climate::terminateCycle() {
    if (this->shouldSendExternalTemperature_) {
        // We will receive an ACK packet for this.
        // Sending WantedSettings must be delayed in this case (lastSend timestamp updated).
        this->sendRemoteTemperature();
    }

    this->loopCycle.cycleEnded();

    if (this->hp_uptime_connection_sensor_ != nullptr) {
        // If the uptime connection sensor is configured,
        // we trigger a manual update at the end of a cycle.
        this->hp_uptime_connection_sensor_->update();
    }

    this->nbCompleteCycles_++;
}
```

**Comment by Dale-Muccignat** [comment]
Unsure if this is specific to this issue, but I observed a similar behaviour with a MSZ-AP25VGD2 where it would revert back to the internal temp after 30 minutes (seemingly precisely) of no change to the remote sensor value. Red being the remote sensor and blue being the current temp reported from the aircon

![Image](https://github.com/user-attachments/assets/70bf726b-5332-4572-9568-766e0ae33b2b)

![Image](https://github.com/user-attachments/assets/b20ddde0-181b-4274-9f01-de0aef39b0b9)

**Comment by anoopr** [comment]
This is happening to me as well on both a MVZ-A30AA7 and SVZ-KP18NA. They are both ducted air handlers. Heartbeat workaround fixed the issue on both heat pumps.

I'll test out the `update_interval` hypothesis and report back.

**Comment by xeoMe015** [comment]
> This is happening to me as well on both a MVZ-A30AA7 and SVZ-KP18NA. They are both ducted air handlers. Heartbeat workaround fixed the issue on both heat pumps.
> 
> I'll test out the `update_interval` hypothesis and report back.


Could you share a daily chart of your temperature history? Maybe one before the fix and one after the fix?
My temperatures go up and down and i'm not exactly sure if this is expected behaviour.

**Comment by luiggi78** [comment]
> Two units seem impacted: a MSZ-GL18NA-U1 and a MSZ-GL06NA-U1

Did you get it to work? I also have MSZ unit.

**Comment by ams2990** [comment]
It took me a few tries to set 102 to 3, but I was able to get it eventually.

**Comment by jimmy-zhening-luo** [comment]
> Also curious if using the heartbeat approach effectively defeats the `remote_temperature_timeout` setting? I used to use that approach with the GeoffDavis implementation, but not sure it's a good idea if my remote sensors fail (i.e. will heartbeat keep synthetically updating the temperature?)

I was thinking about this. @messnerdev , what do you think about something like this?

- Intuitively, this would refuse to emit either `nan` or out of range values.
- I also switched to using a priority throttle also prioritize emitting a `nan`.
- Using a timeout instead of a heartbeat seems more functionally correct -- keep blasting every 15 seconds but allow a real value to interrupt and reset the clock.

That said, logically if your unit is forgetting its remote temperature if not emitted again in 45s, that's already your effective (very bad) remote temperature timeout. All my filter changes do is better ensure that no values are emitted in the nan case for when your sensor is actually down. And, when your sensor is back, it will start emitting correctly again. It also survives things like Home Assistant restarts that happen in fewer than 45 seconds typically. That said, this is only suitable if your sensor is generally stable. And if any reader is seeing this and you have a very noisy sensor than you have the opposite problem so this solution (and this bug) don't apply to you most likely. Anyway here's the filter: 

```yaml
    filters:
      - throttle_with_priority:
          timeout: 1s
          value: [nan]
      - timeout:
          timeout: 15s
          value: last
      - lambda: return (x - 32.0) * (5.0/9.0); # if using °F
      - clamp:
          min_value: 1
          max_value: 40
          ignore_out_of_range: true
      - filter_out: nan
```

**Comment by jimmy-zhening-luo** [comment]
> > I have the same issue, it is updating from the remote sensor, but it's mixing up for some reason. I also noticed my unit is not stopping when at target temperature, there seems to be a built-in offset of at least 1 degree.
> > [@echavet](https://github.com/echavet) any idea in which direction a solution can be found for this issue?
> 
> I have the same my AC has a offset of 1 degree and the current_temperature is set to 0.1 and the target_temperature is set to 0.5 however whenever my remote temperature measures 20.3 it mentions current temperature 20.5 so it’s using steps of 0.5 even though its configured as 0.1
> 
> Setting the AC to set temperature 19 degrees my remote temperature sensor bounces between 19.7 and 20.3. 20.3 makes it 20.5 so the unit stops and 19.7 makes it 19.5 so the unit starts. Not sure if this is normal behaviour?
> 
> I also have the heartbeat configured with 30s and one unit on 15s but same results for both.
> 
> Each few seconds i see cycle start, cycle ended in 0.8 seconds is this expected?
> 
> Outdoor unit: MXZ-3F54
> 
> Indoor units: MSZ-AY42 and MSZ-AY25

I am guessing this is not related to the remote temperature bug. Because I observe this behavior with or without remote temperature. If someone else sees this, does ___anybody___ have units that successfully use currently room temperature deltas smaller than 0.5°C?

**Comment by jimmy-zhening-luo** [comment]
Incidentally, @ams2990 you can probably help with this too since your code also touched this last: I see that both your code (which in Fahrenheit mode uses your mapping) and the original took the final Celsius float and divided it by 2, yielding those half Celsius fluctuations we observe above. Can we therefore assume that the Mitsubishi units indeed can only understand even current room temperature in 0.5°C steps? Remote or not.

I will save my choice words for the genius engineers at Mitsubishi who designed one of the best, most efficient continuous heat pumps known to humankind, capable of maintaining a room temperature perfectly if it wanted to, yet decided to make it worthless by continuously wildly oscillate between two 0.5°C steps.

**Comment by ams2990** [comment]
> I'm digging into all this, I'm also confused why this remote temperature update code was removed from a seemingly unrelated patch: [0e5e805](https://github.com/echavet/MitsubishiCN105ESPHome/commit/0e5e805b7b7cb56640373b44abd7437f6dcca2f7)
> 
> [@ams2990](https://github.com/ams2990) I think you mentioned there was some buggy code you removed. Since you've touched it, I'm wondering if you have any ideas off the top of your head about why remote temperature seems to not stick past 60 seconds?

It wasn't removed, the function just moved from componentEntries.cpp to climateControls.cpp, so it could use the `mapCelsiusForConversionFromFahrenheit` function I added to that file.

I don't have any insight into why the value doesn't stick. I observed this on my own units and worked around it with a 15s heartbeat.


> Incidentally, [@ams2990](https://github.com/ams2990) you can probably help with this too since your code also touched this last: I see that both your code (which in Fahrenheit mode uses your mapping) and the original took the final Celsius float and divided it by 2, yielding those half Celsius fluctuations we observe above. Can we therefore assume that the Mitsubishi units indeed can only understand even current room temperature in 0.5°C steps? Remote or not.

That's correct, if you look at the protocol, the 0.5°C steps are converted into integers when passed over the wire, so there's no room for a value in between.

> I will save my choice words for the genius engineers at Mitsubishi who designed one of the best, most efficient continuous heat pumps known to humankind, capable of maintaining a room temperature perfectly if it wanted to, yet decided to make it worthless by continuously wildly oscillate between two 0.5°C steps.

I also find this frustrating because my thermostats can measure in 0.5°F, but my air handler can do at best 1.0°F. To avoid the oscillation problem, I created a "lagging" sensor that ignores half-degree changes. This is a little different than rounding, because if the sensor moves from 1.0 to 1.5, it stays at 1, or if it moves from 2.0 to 1.5, it stays at 2. This helps reduce oscillation by requiring the temperature to move by a full degree. Here's what that looks like, if you're interested:
```jinja
{% set new =
     states('sensor.2nd_floor_temperature') | float %}
{% if new | round == new %}
  {{ new }}
{% else %}
  {% set old =
       states('sensor.2nd_floor_temperature_lagging') %}
  {% set rounding_mode = 
    'common'
    if old in ('unavailable', 'unknown') 
    else 'ceil'
    if new < old | int 
    else 'floor' %}
  {{ new | round(method=rounding_mode) }}
{% endif %}
```

**Comment by jimmy-zhening-luo** [comment]
I am still reading / processing, but that filter you shared is brilliant and seems very promising to me. Thank you for sharing and for your fast response and help.

**Comment by jimmy-zhening-luo** [comment]
Still wrapping my head around your template sensor; in the meantime I've implemented  it and am watching it, without feeding it to the heat pump yet (brain not ready for those second-order dynamics). For other people interested in quickly testing, here's the full template config version of ams2990's rounded temperature sensor:

```yaml
name: Climate Guest Temperature Rounded
unique_id: guest_temperature_rounded
state_class: measurement
device_class: temperature
unit_of_measurement: °F
availability: "{{ has_value('sensor.climate_guest_temperature') }}"
state: >-
  {% set new = states('sensor.climate_guest_temperature') | float %}
  {{
  new if new == new | round
  else new | round(
    0,
    'common' if not has_value(this.entity_id)
    else
      'ceil' if states(this.entity_id) | int > new
      else 'floor'
  )
  }}


```

**Comment by ams2990** [comment]
> When heating (room temperature below 21.5 °C), round down the temperature => heating will not be terminated prematurely by rounding up; when cooling, round down and do not terminate cooling prematurely.

I used to use a similar strategy, but I observed two problems: 1) The oscillating. A small change in the ambient temperature could move my rounded sensor by a full degree F. The lagging sensor above is a lot more stable, because it requires a larger change in the input to produce a change in the output. 2) It was overcooling. I observed my unit to overshoot with "always round", to around 1.5°F, whereas it stays much closer with my approach above.

**Comment by xeoMe015** [comment]
I think the devices just have a present build-in to cool or heat above the set temperature.
From the debug logs you can also see target -1.0

[14:43:16][D][SETTINGS:105]: [current]-> [power: OFF, target °C: 24.0, mode: HEAT, fan: AUTO, vane: ↑, wvane: ←]
[14:43:16][D][SETTINGS:105]: [received]-> [power: OFF, target °C: 24.0, mode: HEAT, fan: AUTO, vane: ↑, wvane: ←]
[14:43:16][D][EVT_SETS:062]: [wanted]-> [power: -, **target °C: -1.0,** mode: -, fan: -, vane: -, wvane: -, hasChanged ? ->  NO, hasBeenSent ? ->  NO]
[14:43:16][D][SETTINGS:089]: [climate]-> [mode: OFF, target °C: 24.0, fan: AUTO, swing: OFF]

My unit allows 0.5 degrees steps in target temperature via this MitsubishiCN105ESPHome project and also via MELCloud however the remote only allows 1 degree steps.
Also oversized units might go to their lowest available kwh ouput but if this is still to much the room overshoots in heating/cooling where it then stops/idles above the target if crossed by a specific range.

I use a thottle_average and with this when cooling is set to 21 my units stops when the remote temperature sensor has reached 19.7 (which then gets rounded by the heatpump to 19.5 and the unit continues the cooling when the remote temperature sensor goes to 20.3 gets arounded again to 20.5

**Comment by disruptivepatternmaterial** [comment]
I think if you look at the code I shared in the discussion about a dual set point you will find (if I understand!) that the meta climate accounts for all this. You can set headbands, overruns, max/min run times, etc... 

https://github.com/echavet/MitsubishiCN105ESPHome/discussions/384

I have a couple more change to put in it, like some changes to the mapping of the modes, but other than that it VERY much made a difference in how these operate that I love. It is worth checking out...

**Comment by ams2990** [comment]
@disruptivepatternmaterial wow, that is excellent! I've got a ton of automation in HA this could conceivably replace. I also like that I could write the automation in C++, which is far more enjoyable and maintainable than HN templates. I'll have to look into all of those settings you're configuring to understand how they work and can be applied to my own setup.

**Comment by echavet** [comment]
> I setup a remote temperature sensor which is correctly updates the temperature based on the HA entity id.  However, the Current Temperate is regularly overwritten by the room temperature being provided by unit.  Note: this behavior is happening on only 1 of my 2 heat pumps (they are different models) configured with a remote thermostat.
> 
> 
> 
> I enabled DEBUG log and captured when it's happening.  In this example, the remote temp is reading 19.5C and the unit is reading 11C.
> 
> 
> 
> ```
> 
> [13:30:27][D][CN105:308]: sending a request for settings packet (0x02)
> 
> [13:30:27][I][CYCLE:039]: 1: Cycle start
> 
> [13:30:27][D][CYCLE:311]: 2a: Sending settings request (0x02)
> 
> [13:30:27][D][CN105:323]: creating Info packet
> 
> [13:30:27][D][CN105:084]: writing packet...
> 
> [13:30:27][D][WRITE:170]: FC 42 01 30 10 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7B 
> 
> [13:30:27][D][Header:091]: command: (62) data length: [10]<-- header
> 
> [13:30:27][D][READ:170]: FC 62 01 30 10 02 00 00 01 01 0F 02 05 00 00 00 A0 00 00 00 00 A3 
> 
> [13:30:27][D][chkSum:077]: OK-> A3=A3 
> 
> [13:30:27][D][CYCLE:315]: 2b: Receiving settings response
> 
> [13:30:27][D][Decoder:173]: [0x02 is settings]
> 
> [13:30:27][D][Decoder:181]: [Power : ON]
> 
> [13:30:27][D][Decoder:182]: [iSee  : 0]
> 
> [13:30:27][D][Decoder:183]: [Mode  : HEAT]
> 
> [13:30:27][D][Decoder:195]: [Temp °C: 16.000000]
> 
> [13:30:27][D][Decoder:198]: [Fan: 1]
> 
> [13:30:27][D][Decoder:201]: [Vane: ↓↓]
> 
> [13:30:27][W][lookup:204]: wideVane reading caution: value 0 not found, returning value at index 0
> 
> [13:30:27][D][Decoder:206]: [wideVane: ←← (adj:0)]
> 
> [13:30:27][D][SETTINGS:113]: [current        ]-> [power: ON , target °C: 16.0, mode: HEAT  , fan: 1     , vane: ↓↓, wvane: ←←]
> 
> [13:30:27][D][SETTINGS:113]: [received       ]-> [power: ON , target °C: 16.0, mode: HEAT  , fan: 1     , vane: ↓↓, wvane: ←←]
> 
> [13:30:27][D][EVT_SETS:072]: [wanted         ]-> [power: -  , target °C: -1.0, mode: -     , fan: -     , vane: -     , wvane: -     , hasChanged ? ->  NO, hasBeenSent ? ->  NO]
> 
> [13:30:27][D][SETTINGS:087]: [climate]-> [mode: HEAT, target °C: 16.0, fan: LOW, swing: OFF]
> 
> [13:30:27][D][CN105:218]: setting action to -> 4
> 
> [13:30:27][D][CN105:267]: Climate mode is: 3
> 
> [13:30:27][D][CN105:268]: Climate action is: 4
> 
> [13:30:27][D][climate:396]: 'Heat Pump - Living Room' - Sending state:
> 
> [13:30:27][D][climate:399]:   Mode: HEAT
> 
> [13:30:27][D][climate:401]:   Action: IDLE
> 
> [13:30:27][D][climate:404]:   Fan Mode: LOW
> 
> [13:30:27][D][climate:416]:   Swing Mode: OFF
> 
> [13:30:27][D][climate:419]:   Current Temperature: 19.50°C
> 
> [13:30:27][D][climate:425]:   Target Temperature: 16.00°C
> 
> [13:30:27][D][CYCLE:318]: 3a: Sending room °C request (0x03)
> 
> [13:30:27][D][CN105:323]: creating Info packet
> 
> [13:30:27][D][CN105:084]: writing packet...
> 
> [13:30:27][D][WRITE:170]: FC 42 01 30 10 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 7A 
> 
> [13:30:27][W][component:237]: Component esphome.coroutine took a long time for an operation (75 ms).
> 
> [13:30:27][W][component:238]: Components should block for at most 30 ms.
> 
> [13:30:27][D][Header:091]: command: (62) data length: [10]<-- header
> 
> [13:30:27][D][READ:170]: FC 62 01 30 10 03 00 00 01 00 92 96 00 00 00 00 00 00 00 00 00 31 
> 
> [13:30:27][D][chkSum:077]: OK-> 31=31 
> 
> [13:30:27][D][CYCLE:324]: 3b: Receiving room °C response
> 
> [13:30:27][D][Decoder:253]: [Room °C: 11.000000]
> 
> [13:30:27][D][Decoder:254]: [OAT  °C: 9.000000]
> 
> [13:30:27][I][STATUS:134]: [received       ]-> [room C°: 11.0, operating: NO , compressor freq: 0.0 Hz]
> 
> [13:30:27][I][STATUS:134]: [current        ]-> [room C°: 19.5, operating: NO , compressor freq: 0.0 Hz]
> 
> [13:30:27][D][CN105:218]: setting action to -> 4
> 
> [13:30:27][D][CN105:267]: Climate mode is: 3
> 
> [13:30:27][D][CN105:268]: Climate action is: 4
> 
> [13:30:27][D][climate:396]: 'Heat Pump - Living Room' - Sending state:
> 
> [13:30:27][D][climate:399]:   Mode: HEAT
> 
> [13:30:27][D][climate:401]:   Action: IDLE
> 
> [13:30:27][D][climate:404]:   Fan Mode: LOW
> 
> [13:30:27][D][climate:416]:   Swing Mode: OFF
> 
> [13:30:27][D][climate:419]:   Current Temperature: 11.00°C
> 
> [13:30:27][D][climate:425]:   Target Temperature: 16.00°C
> 
> [13:30:27][D][CYCLE:327]: 4a: Sending status request (0x06)
> 
> [13:30:27][D][CN105:323]: creating Info packet
> 
> [13:30:27][D][CN105:084]: writing packet...
> 
> [13:30:27][D][WRITE:170]: FC 42 01 30 10 06 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 77 
> 
> [13:30:27][W][component:237]: Component esphome.coroutine took a long time for an operation (56 ms).
> 
> [13:30:27][W][component:238]: Components should block for at most 30 ms.
> 
> [13:30:27][D][Header:091]: command: (62) data length: [10]<-- header
> 
> ```
> 
> Here is the relevant parts of the config yaml:
> 
> ```yaml
> 
>  sensor:
> 
>    - platform: homeassistant
> 
>      name: "Temperature Sensor From Home Assistant"
> 
>      entity_id: ${remote_temp_sensor}    
> 
>      internal: false
> 
>      device_class: temperature
> 
>      state_class: measurement
> 
>      unit_of_measurement: "°C"
> 
>      filters:
> 
>        - lambda: 'return (x - 32) * (5.0/9.0);'
> 
>        - clamp: # Limits values to range accepted by Mitsubishi units
> 
>            min_value: 1
> 
>            max_value: 40
> 
>            ignore_out_of_range: true
> 
>        - throttle: 30s
> 
>      on_value:
> 
>        then:
> 
>          - logger.log:
> 
>              level: WARN
> 
>              format: "Remote temperature received from HA: %.1f C"
> 
>              args: [ 'x' ]
> 
>          - lambda: 'id(hp).set_remote_temperature(x);'
> 
> 
> 
>  climate:
> 
>    - platform: cn105
> 
>      name: ${friendly_name}
> 
>      id: hp
> 
>      remote_temperature_timeout: 30min
> 
>      update_interval: 4s
> 
>      debounce_delay : 100ms
> 
> ```
> 
> 
> 
> Lastly, this is easily reproducible on my side.  It oscillates between the two temp readings every few seconds, happy to provide any more info if it'd be helpful.

Hi,
Did you find a cause for your issue?
If not, can you tell me more about the sensor you're using as ${remote_temp_sensor}.
I have faced the same issue yesterday with a zigbee thermometer which was very lazy with sending sensor updates. And it looks like my Mitsubishi unit ends up toggling to internal sensor when no external temperature is sent.

**Comment by echavet** [comment]
> I had similar issues when relying on the update frequency of zigbee sensors. The controller seems to like frequent updates, even if the value does not change.

Yes I have exactly the same issue.
I has to set the timeout to 1 hour and to change the sonoff SNZB-02D reports frequency.

Max Rep Interval and Min Rep Change parameters can be set from zigbee2MQTT report section for the device. 
You have to press the back button on the device to allow it to listen to z2mqtt change requests. 

"Min Rep change" must be set to a smaller value : 10 or 20 for 0.1 or 0.2 range.

**Comment by kchen** [comment]
I don't know if this is the case for anyone else here, but when I initially encountered this problem, I had connected my ESP32 to my PAC-USWHS002-WF-2 (and removed the WF-2's remote sensor in the Mitsubishi Comfort app), and the WF-2 was then connected to my indoor unit. I tried several of the workarounds in this thread, and they failed to fix the problem.

However, once I removed the WF-2 and connected the ESP32 directly to the indoor unit's CN105 port, I believe the remote temperature sensor defined in the ESP32 started working properly for me.

I assume that once I removed the WF-2's remote sensor, the WF-2 may have started instructing the unit to use its internal sensor, so even if I attempted to use the "Set Functions" workaround in this ticket, the WF-2 may have still been overwriting that setting.

**Comment by ams2990** [comment]
Not the case for me -- I had this problem with direct connection to the indoor unit.

**Comment by raiseandfall** [comment]
@rudizl you can add the following config to your `climate`: https://github.com/echavet/MitsubishiCN105ESPHome/issues/204#issuecomment-2973215545

**Comment by glarizza** [comment]
Just want to add that I also hit the same problem as echavet describes here: https://github.com/echavet/MitsubishiCN105ESPHome/issues/204#issuecomment-3620372922 with the same Sonoff SNZB-02 Temperature sensor and Zigbee2mqtt:

<img width="613" height="354" alt="Image" src="https://github.com/user-attachments/assets/b69a5575-a8cb-414e-bfed-b5930dc3ed4e" />
 
<img width="1234" height="259" alt="Image" src="https://github.com/user-attachments/assets/448b4024-4e29-494f-b70d-f408aa478afe" />

 The default `max_rep_interval` value for that device in Zigbee2mqtt  is 1800 seconds (30 minutes) which is the same default value as the `remote_temperature_timeout` esphome setting. So if esphome doesn't get a value in that time period then it fails over to the internal thermometer setting. Either decreasing the `max_rep_interval` for the device or increasing the `remote_temperature_timeout` (or both) solved this problem for me.

**Comment by echavet** [comment]
Hey I know this pb too.
I had to switch to a SNZB-02P because my SNZB-02D had a too few temp reports with zigbee2mqtt.


The Sonoff SNZB-02D sensors have hardcoded firmware that ignores standard Zigbee reporting configurations via Zigbee2MQTT. Specifically, they do not respect the Max rep interval (Maximum Reporting Interval) command. To preserve battery life, the sensor only transmits data when it detects a significant change in temperature or humidity. If the environment is stable, the sensor can remain silent for over an hour. As a result, Zigbee2MQTT and connected systems (like Home Assistant or ESPHome) often mark the device as unavailable due to the lack of regular heartbeat updates.

What sensor are you using?

**Comment by 01awolf** [comment]
Hi @echavet I have also the same sensors SNZB-02P.
My Max rep interval is set to 3600 so I changed my remote_temperature_timeout to 65 mins.
Let's see if it's correcting the issue.

BTW great work ! Cocorico !

**Comment by echavet** [comment]
In addition to ignoring the maximum reporting interval, the SNZB-02D sensor also fails to respect the Min rep change (Minimum Reporting Change) configuration in Zigbee2MQTT. This parameter is designed to let users define the exact temperature or humidity variation (e.g., 0.1°C or 0.2°C) required to trigger a new data transmission. However, the device typically relies on its own hardcoded factory thresholds to maximize battery life. Because it ignores these custom sensitivity settings, the sensor remains completely silent during periods of slow temperature variation or perfect environmental stability, creating problematic data gaps.

**Comment by 01awolf** [comment]
@echavet could you share with us your SNZB-02P configuration ?
And your remote_temperature_timeout.

BTW, do you know what side effect this workaround could have ?
filters:
    - heartbeat: 15s 

Thanks !

**Comment by echavet** [comment]
![Screenshot_20260503-195348.png](https://github.com/user-attachments/assets/875e2376-5157-406e-ad24-1c5b39e460fe)

Groupe de fonctionnalités *
msTemperatureMeasurement
Attribut *
measuredValue
Intervalle min de rapport *
10
Intervalle max de rapport *
1800
Changement min de rapport *
0

**Comment by echavet** [comment]
You need to press the button on the device so as it records the new values

**Comment by 01awolf** [comment]
Thank you, @echavet.
The only difference in my setup is that the 'Minimum report change' is set to 100.

According to this link below, the report should still be sent after 'Max rep interval' regardless of the 'Minimum report change' parameter ?
https://julien-moreau.fr/2025/07/17/z2m-les-capteurs-de-temperature/

Observations :

Min rep interval : 10 [secondes]
Max rep interval 6500 [secondes]
Min rep change 100 [1/100°C] <= unité de précision : 0.01°C
En clair nous avons les instructions suivantes :

Pas plus de 1 rapport toutes les 10 secondes si le seuil des 1°C [100 fois 1/100ème] et franchi à chaque fois mais envoyer un rapport systématique toutes les 1h48 et 33secondes [6500secondes] **si ce seuil n’est pas franchi.**

**Comment by echavet** [comment]
Tout ce que je peux dire c'est que le précédent capteur que j'utilisais avait une sensibilité inférieure. De mémoire de 0.2° alors que celui-ci a une sensibilité de 0.1.
Cela a suffit dans ma conf à m'éviter une bascule sur la sonde interne lorsque la température de consigne est atteinte.
Ces appareils ne respectent pas parfaitement les paramètres a cause de leurs contraintes d'autonomie.
Je n'ai pas trouvé de meilleure solution que de placer le capteur en dehors du flux, d'augmenter la durée du timeout à 90 minutes, et de positionner les paramètres de la sonde a la plus grande sensibilité avec des délais le plus courts possible.
Depuis je n'ai pas supprimé totalement les bascules, mais elles se produisent rarement.
Je ne crois pas qu'il y ait de solution. Si le capteur ne veut pas parler pendant une durée>90 minutes ou si le split a un timeout interne, la sonde bascule.

**Comment by niobos** [comment]
Hi,
I'm experiencing something similar to what is reported in this issue: The unit returns to use its own internal temperature sometimes. And I agree with what has been mentioned above. I think the problem is a "too stable" external temperature reading:

Below, you can find the "room temperature" as reported by the A/C unit in blue, and the HA-sensor that is set as remote temperature in red.

<img width="1227" height="374" alt="Image" src="https://github.com/user-attachments/assets/ecb0906f-3dcc-4edb-af0d-e81844b25e9e" />

<img width="1231" height="364" alt="Image" src="https://github.com/user-attachments/assets/772998b5-bc22-4917-baee-48c0b7b4b519" />

What I notice is that the A/C unit returns to the internal temperature after a long stable external temperature without change. During periods with (minor) changes, it returned to use the external temp, and kept using the external temp.

I think the `remote_temperature_timeout` (set to 5 minutes in my case) is falsely trigger when the sensor simply doesn't change. I know the HA suppresses updates when nothing changes, so it might be difficult to differentiate between "no more updates received from HA" versus "still same temperature".

Adding a heartbeat in the ESPHome code "solves" the issue, but effectively disables the timeout functionality.

**Comment by echavet** [comment]
Yes.
I went through this also. Found no other way but changing the remote temperature sensor for a more sensitive one.

**Comment by JulianGroeli** [comment]
New data point here since my setup is different from what's been discussed. Instead of tweaking the heartbeat/timeout settings on the existing remote-temp pipeline, I used Node Red and MQTT to poll the remote temp sensor and push the value to ESPHome.  Issue resolved over a full day/night of testing, No other changes.

**Setup:** D1 mini / ESP8266, MLZ-KP18NA indoor unit, remote temp sourced from an Aqara Zigbee sensor via Zigbee2MQTT → HA. Originally used the standard `platform: homeassistant` sensor + `remote_temperature_source:` pattern with `remote_temperature_timeout: 30min` and `remote_temperature_keepalive_interval: 20s`. Saw the same intermittent fallback-to-internal-sensor behavior described throughout this thread — current temperature would jump several degrees above/below the real remote reading for anywhere from several minutes to ~30 min, then self-correct, recurring multiple times per day with no fixed cadence.

Ruled out before testing further:
- Zigbee/HA entity availability — confirmed via history that the source HA entity never went `unavailable` during spike windows
- WiFi roam-scan disruption — found and fixed a separate issue where ESPHome's `post_connect_roaming` (default `true`) was causing periodic brief WiFi disconnects on this ESP8266; setting `post_connect_roaming: false` fixed *that* issue but didn't stop the temperature fallback
- API/HA connection drops — confirmed via HA's own `aioesphomeapi` debug logs that the connection stayed live and connected through several observed spike events

**Change tested:** replaced the `platform: homeassistant` sensor with a `platform: mqtt_subscribe` sensor, fed by an external Node-RED flow that polls the HA entity every 30s and republishes to a plain MQTT topic regardless of whether the value changed (i.e. a continuous heartbeat-style push, external to ESPHome). `climate:` config (including `remote_temperature_source`, timeout, and keepalive settings) was left completely unchanged — only the delivery mechanism for the underlying sensor value changed.

**Result:** zero fallback events over 120h+ of testing (previously multiple per day), with the sensor otherwise behaving identically. Graphs attached (before/after).

This doesn't pin down the exact internal cause, but suggests the event-driven nature of the native `platform: homeassistant` sensor (only pushing on real HA state-change events) may be a meaningful factor distinct from Zigbee sensor silence — since in this case the source entity was never observed unavailable or stale, yet switching to an externally-driven, continuously-republishing MQTT feed with the exact same climate-side config resolved it.

Happy to share the full YAML or debug logs if useful.

Before:
<img width="3315" height="644" alt="Image" src="https://github.com/user-attachments/assets/86900787-960d-4d64-a6c4-5ccab734579a" />
Red arrows = fallback spikes, yellow arrows = fallback drops, dashed blue lines = disconnect/reconnect events (confirmed unrelated to the spikes, per the "Using Remote Temperature" sensor and connection logs).


After:
<img width="3559" height="492" alt="Image" src="https://github.com/user-attachments/assets/a56450a2-54d9-4878-b765-90bf28c50843" />

Note tight alignment between remote temp and reported temp

**Comment by 01awolf** [comment]
Thank you, Julian. This is really interesting !
I'm not familiar with Node-RED.
Would you be able to write a tutorial or create a video showing the exact steps to implement this workaround ?
Thank you !

**Comment by niobos** [comment]
> I'm not familiar with Node-RED. Would you be able to write a tutorial or create a video showing the exact steps to implement this workaround ? Thank you !

Although Node-Red is great, you can do this in pure Home Assistant as well:

1. Add a Home Assistant service to the ESPHome config for your devices as [documented here](https://github.com/echavet/MitsubishiCN105ESPHome#alternate---homeassistant-action)
2. Create an automation that runs on a Time Pattern. Pick a pattern that suits your need, and is shorter than your fallback timeout. I use "every minute" (i.e. `*` in the Minutes field, nothing in Hours and Seconds)
3. Use the `set_remote_temperature` action you just added in step (1) for each A/C you want to use remote temperature on. Specify what temperature sensor to use in the data (e.g. `{{ states('sensor.living_room_temperature')|float }}`)

**Comment by JulianGroeli** [comment]
@01awolf If you don't already have MQTT or Node-RED set up, @niobos's suggestion above (pure HA automation + `time_pattern` trigger calling the `set_remote_temperature` action) is the simpler starting point. No new infrastructure required, just an ESPHome config change and one HA automation. Worth trying that first.

If you want to learn about MQTT/Node-RED generally (useful for lots of other things too, not just this):
- MQTT + the Mosquitto broker add-on: [Home Assistant's MQTT integration docs](https://www.home-assistant.io/integrations/mqtt/)
- Node-RED: [official Node-RED docs](https://nodered.org/docs/) and the [Node-RED HA add-on](https://github.com/hassio-addons/addon-node-red)

If you already have MQTT running (I was using it for Zigbee2MQTT and the prior `gysmo38/mitsubishi2MQTT` implementation), here's the relevant YAML block. It replaces the native `platform: homeassistant` sensor with one fed by an external process (Node-RED, or any other MQTT publisher) that republishes on a fixed interval regardless of whether the value changed. Same "heartbeat" principle as the workarounds already discussed, just implemented at the delivery layer instead of a sensor filter.

```yaml
mqtt:
  broker: !secret mqtt_broker
  username: !secret mqtt_username
  password: !secret mqtt_password
  discovery: false   # important — see note below

sensor:
  - platform: mqtt_subscribe
    id: ha_remote_temp
    topic: esphome/<your-device>/remote_temp
    internal: true
    # filters:
    # - lambda: return (x - 32) * (5.0/9.0);   # convert unit if needed

climate:
  - platform: cn105
    # ...your existing climate config...
    remote_temperature_source:
      sensor_id: ha_remote_temp
```

Then just have something (Node-RED, an automation, whatever) publish the current temperature as a plain number to that topic on a fixed interval - I use 30s.

One gotcha worth flagging: `discovery: false` needs to be there from the very first time you add the `mqtt:` block, or ESPHome will auto-publish HA-discovery messages and you'll end up with duplicate entities that are annoying to clean up afterward.

Since both approaches work by *actively re-pushing* the temperature on a fixed schedule from outside the `remote_temperature_source` binding, rather than relying on it to pull from a passively-subscribed HA sensor. If @niobos's method holds up the same way mine has, that convergence would point toward the bug being in how the native `platform: homeassistant` sensor delivers updates (or how the component's fallback logic interacts with it), rather than something inherent to the sensor hardware itself, since (I believe) neither of us changed our physical sensor at all.

**Comment by echavet** [comment]
## Maintainer update / roadmap

Still **P1**: remote temperature must not be overwritten by the indoor unit’s local reading when remote temp is active.

We will treat this carefully (multi-model): any fix should be gated on “remote temperature mode engaged”, with clear keepalive/timeout behaviour — not a blanket ignore of all room-temp frames for every install.

If you can still reproduce on current `main`, a short log with remote temp enabled showing the overwrite (Decoder room °C vs remote) would help prioritise the PR.
